File: core/src/main/java/gg/norisk/core/manager/PayloadManager.java
package gg.norisk.core.manager;

import gg.norisk.core.channel.ChannelApi;
import gg.norisk.core.payloads.AbstractPayload;

import java.util.Timer;
import java.util.TimerTask;
import java.util.UUID;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ConcurrentLinkedQueue;
import java.util.function.BiConsumer;

public class PayloadManager {
    private final ConcurrentHashMap<UUID, ConcurrentLinkedQueue<AbstractPayload>> playerPayloads = new ConcurrentHashMap<>();
    private final ConcurrentHashMap<String, Class<? extends AbstractPayload>> registeredPayloads = new ConcurrentHashMap<>();
    private final ConcurrentHashMap<UUID, Long> payloadSendAllowedAt = new ConcurrentHashMap<>();
    private final ConcurrentHashMap<String, Timer> payloadTimeouts = new ConcurrentHashMap<>();

    private final long PAYLOAD_DELAY_MS = 5000L;
    private final long PAYLOAD_TIMEOUT_MS = 2000L;

    private final ChannelApi channelApi;

    public PayloadManager() {
        this.channelApi = new ChannelApi();
    }

    private BiConsumer<UUID, byte[]> sendToClientCallback;

    public void registerPayloadType(String type, Class<? extends AbstractPayload> clazz) {
        registeredPayloads.put(type, clazz);
    }

    public void registerPlayer(UUID uuid) {
        playerPayloads.putIfAbsent(uuid, new ConcurrentLinkedQueue<>());
    }

    public void unregisterPlayer(UUID uuid) {
        playerPayloads.remove(uuid);
    }

    public void queuePayload(UUID uuid, AbstractPayload payload) {
        playerPayloads.computeIfAbsent(uuid, k -> new ConcurrentLinkedQueue<>()).add(payload);
    }

    public void sendToPlayer(UUID uuid, AbstractPayload payload, BiConsumer<UUID, byte[]> sendToClient) {
        Long allowedAt = payloadSendAllowedAt.get(uuid);
        if (allowedAt != null && System.currentTimeMillis() < allowedAt) {
            queuePayload(uuid, payload);
            return;
        }

        byte[] data = channelApi.send(payload);
        sendToClient.accept(uuid, data);

        String key = uuid + ":" + payload.getType();
        Timer existingTimer = payloadTimeouts.remove(key);
        if (existingTimer != null) {
            existingTimer.cancel();
        }

        Timer timer = new Timer();
        timer.schedule(new TimerTask() {
            @Override
            public void run() {
                payloadTimeouts.remove(key);
            }
        }, PAYLOAD_TIMEOUT_MS);
        payloadTimeouts.put(key, timer);
    }

    public void onPayloadAck(UUID uuid, String payloadType) {
        String key = uuid + ":" + payloadType;
        Timer timer = payloadTimeouts.remove(key);
        if (timer != null) {
            timer.cancel();
        }
    }

    public void sendQueuedPayloads(UUID uuid, BiConsumer<UUID, byte[]> sendToClient) {
        Long allowedAt = payloadSendAllowedAt.get(uuid);
        if (allowedAt != null && System.currentTimeMillis() < allowedAt) {
            return;
        }

        ConcurrentLinkedQueue<AbstractPayload> queue = playerPayloads.get(uuid);
        if (queue == null) return;

        while (!queue.isEmpty()) {
            AbstractPayload payload = queue.poll();
            if (payload != null) {
                sendToPlayer(uuid, payload, sendToClient);
            }
        }
    }

    public void broadcast(AbstractPayload payload, BiConsumer<UUID, byte[]> sendToClient) {
        for (UUID uuid : playerPayloads.keySet()) {
            sendToPlayer(uuid, payload, sendToClient);
        }
    }

    public void setSendToClientCallback(BiConsumer<UUID, byte[]> callback) {
        sendToClientCallback = callback;
    }

    public void onPayloadReceived(UUID sender, String payloadJson) {
        if (payloadJson.contains("\"type\":\"handshake\"")) {
            payloadSendAllowedAt.put(sender, System.currentTimeMillis() + PAYLOAD_DELAY_MS);

            Thread thread = new Thread(() -> {
                try {
                    Thread.sleep(PAYLOAD_DELAY_MS);
                    payloadSendAllowedAt.remove(sender);
                    BiConsumer<UUID, byte[]> callback = sendToClientCallback;
                    if (callback != null) {
                        sendQueuedPayloads(sender, callback);
                    }
                } catch (InterruptedException e) {
                    Thread.currentThread().interrupt();
                }
            });
            thread.start();
        }
    }

    public void onPlayerLeave(UUID uuid) {
        // Remove all payload timeouts for this player
        payloadTimeouts.entrySet().removeIf(entry -> {
            String key = entry.getKey();
            if (key.startsWith(uuid.toString() + ":")) {
                entry.getValue().cancel();
                return true;
            }
            return false;
        });
    }
} 

File: core/src/main/kotlin/gg/norisk/core/manager/PayloadManager.kt
package gg.norisk.core.manager

import gg.norisk.core.payloads.AbstractPayload
import java.util.UUID
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.ConcurrentLinkedQueue

object PayloadManager {
    private val playerPayloads = ConcurrentHashMap<UUID, ConcurrentLinkedQueue<AbstractPayload>>()
    private val registeredPayloads = ConcurrentHashMap<String, Class<out AbstractPayload>>()
    private val payloadSendAllowedAt = ConcurrentHashMap<UUID, Long>()

    private val payloadTimeouts = ConcurrentHashMap<Pair<UUID, String>, java.util.Timer>()

    private const val PAYLOAD_DELAY_MS = 5000L
    private const val PAYLOAD_TIMEOUT_MS = 2000L

    private var sendToClientCallback: ((UUID, ByteArray) -> Unit)? = null

    fun registerPayloadType(type: String, clazz: Class<out AbstractPayload>) {
        registeredPayloads[type] = clazz
    }

    fun registerPlayer(uuid: UUID) {
        playerPayloads.putIfAbsent(uuid, ConcurrentLinkedQueue())
    }

    fun unregisterPlayer(uuid: UUID) {
        playerPayloads.remove(uuid)
    }

    fun queuePayload(uuid: UUID, payload: AbstractPayload) {
        playerPayloads.computeIfAbsent(uuid) { ConcurrentLinkedQueue() }.add(payload)
    }

    fun sendToPlayer(uuid: UUID, payload: AbstractPayload, sendToClient: (UUID, ByteArray) -> Unit) {
        val allowedAt = payloadSendAllowedAt[uuid]
        if (allowedAt != null && System.currentTimeMillis() < allowedAt) {
            queuePayload(uuid, payload)
            return
        }
        val data = ChannelApi.send(payload)
        sendToClient(uuid, data)
        val key = uuid to payload.type
        payloadTimeouts[key]?.cancel()
        val timer = java.util.Timer()
        timer.schedule(object : java.util.TimerTask() {
            override fun run() {
                payloadTimeouts.remove(key)
            }
        }, PAYLOAD_TIMEOUT_MS)
        payloadTimeouts[key] = timer
    }

    fun onPayloadAck(uuid: UUID, payloadType: String) {
        val key = uuid to payloadType
        payloadTimeouts.remove(key)?.cancel()
    }

    fun sendQueuedPayloads(uuid: UUID, sendToClient: (UUID, ByteArray) -> Unit) {
        val allowedAt = payloadSendAllowedAt[uuid]
        if (allowedAt != null && System.currentTimeMillis() < allowedAt) {
            return
        }
        val queue = playerPayloads[uuid] ?: return
        while (queue.isNotEmpty()) {
            val payload = queue.poll()
            if (payload != null) sendToPlayer(uuid, payload, sendToClient)
        }
    }

    fun broadcast(payload: AbstractPayload, sendToClient: (UUID, ByteArray) -> Unit) {
        playerPayloads.keys.forEach { uuid ->
            sendToPlayer(uuid, payload, sendToClient)
        }
    }

    fun setSendToClientCallback(callback: (UUID, ByteArray) -> Unit) {
        sendToClientCallback = callback
    }

    fun onPayloadReceived(sender: UUID, payloadJson: String) {
        if (payloadJson.contains("\"type\":\"handshake\"")) {
            payloadSendAllowedAt[sender] = System.currentTimeMillis() + PAYLOAD_DELAY_MS
            val uuid = sender
            Thread {
                Thread.sleep(PAYLOAD_DELAY_MS)
                payloadSendAllowedAt.remove(uuid)
                val callback = sendToClientCallback
                if (callback != null) {
                    sendQueuedPayloads(uuid, callback)
                } else {
                }
            }.start()
        }
    }

    fun onPlayerLeave(uuid: UUID) {
        payloadTimeouts.keys.filter { it.first == uuid }.forEach {
            payloadTimeouts.remove(it)?.cancel()
        }
    }
}